_pytest/mark.py

```diff
@@ -329,30 +336,40 @@ def __call__(self, *args, **kwargs):
             is_class = inspect.isclass(func)
             if len(args) == 1 and (istestfunc(func) or is_class):
                 if is_class:
-                    if hasattr(func, 'pytestmark'):
-                        mark_list = func.pytestmark
-                        if not isinstance(mark_list, list):
-                            mark_list = [mark_list]
-                        # always work on a copy to avoid updating pytestmark
-                        # from a superclass by accident
-                        mark_list = mark_list + [self]
-                        func.pytestmark = mark_list
-                    else:
-                        func.pytestmark = [self]
+                    apply_mark(func, self.mark)
                 else:
-                    holder = getattr(func, self.name, None)
-                    if holder is None:
-                        holder = MarkInfo(self.mark)
-                        setattr(func, self.name, holder)
-                    else:
-                        holder.add_mark(self.mark)
+                    apply_legacy_mark(func, self.mark)
+                    apply_mark(func, self.mark)
                 return func
 
         mark = Mark(self.name, args, kwargs)
         return self.__class__(self.mark.combined_with(mark))
 
 
+def apply_mark(obj, mark):
+    assert isinstance(mark, Mark), mark
+    """applies a marker to an object,
```

Comment:
I had to review the PR, the code, the tests and use the debugger to finally get a picture of what's going on. 😅 

OK IIUC `apply_mark` only sets or updates the `pytestmark` attribute of the given `obj`, while `apply_legacy_mark` actually adds/updates the `MarkInfo` instance with the mark name in `obj`.

Perhaps the docstring of both methods should describe exactly what's going on? Suggestions:

```python
def apply_mark(obj, mark):
    """
    Sets or updates the ``pytestmark`` attribute on ``obj`` with the given ``mark`` object.

    In the future (see issue X), this is the only effect ``@pytest.mark`` will have, and the
    mark transfering from classes and modules to test functions will be done on 
    ``pytest.Item`` objects instead.
    """


def update_legacy_mark_info(obj, mark):
    """
    Updates or creates a ``MarkInfo`` as an attribute of the given object.

    This is how markers are transfered from classes and modules to test functions in the legacy way,
    which is planned to be phased out.
    """
```

Also, we are now introducing a public way to get "pristine" (non-inherited) markers on functions by using a machanism which is already used elsewhere, the `pytestmark` attribute. The `pytestmark` attribute in classes and modules has a different semantic in that they will transfer their marks to child test items.

I suggest we take this opportunity to introduce a proper public API which hides this detail, like `pytest.get_pristine_marks`:

```python
def get_pristine_marks(obj):
    """
    Returns pytest marks applied directly to the given object, without any inheritance from
    parent classes decorated with ``@pytest.mark`` or parent modules with the ``pytestmark`` 
    attribute.
    """
```

Then we don't have to "promise" that we will be supporting a `pytestmark` attribute in function objects directly in the future. If this suggesiton sounds good, I also believe we should name the new `pytestmark` in functions to `_pytest_marks` to avoid users using it by accident.

On the other hand now that I think about it, `pytestmark` in modules have to be implemented like this given that you can't really decorate a module, so perhaps in the end `pytestmark` as the holder for pristine marks is the way to go anyway.


What do you think?